src/hashset.c

```diff
@@ -0,0 +1,1681 @@
+ * to trigger rehashing and cleaning up tombstones. */
+static int cleanUpTombstonesIfNeeded(hashset *s) {
+    if (hashsetIsRehashing(s) || resize_policy == HASHSET_RESIZE_FORBID) {
+        return 0;
+    }
+    if (s->everfulls[0] * 100 >= numBuckets(s->bucketExp[0]) * MAX_FILL_PERCENT_SOFT) {
+        return resize(s, s->used[0], NULL);
+    }
+    return 0;
+}
+
+/* Returns 1 if the table is expanded, 0 if not expanded. If 0 is returned and
+ * 'malloc_failed' is proveded, it is set to 1 if malloc failed and 0
+ * otherwise. */
+static int expand(hashset *s, size_t size, int *malloc_failed) {
+    if (size < hashsetSize(s)) {
+        return 0;
+    }
+    return resize(s, size, malloc_failed);
+}
+
+/* Finds an element matching the key. If a match is found, returns a pointer to
+ * the bucket containing the matching element and points 'pos_in_bucket' to the
+ * index within the bucket. Returns NULL if no matching element was found.
+ *
+ * If 'table_index' is provided, it is set to the index of the table (0 or 1)
+ * the returned bucket belongs to. */
+static bucket *findBucket(hashset *s, uint64_t hash, const void *key, int *pos_in_bucket, int *table_index) {
+    if (hashsetSize(s) == 0) return 0;
+    uint8_t h2 = highBits(hash);
+    int table;
+
+    /* Do some incremental rehashing. */
+    rehashStepOnReadIfNeeded(s);
+
+    /* Check rehashing destination table first, since it is newer and typically
+     * has less 'everfull' flagged buckets. Therefore it needs less probing for
+     * lookup. */
+    for (table = 1; table >= 0; table--) {
+        if (s->used[table] == 0) continue;
+        size_t mask = expToMask(s->bucketExp[table]);
+        size_t bucket_idx = hash & mask;
+        size_t start_bucket_idx = bucket_idx;
+        while (1) {
+            bucket *b = &s->tables[table][bucket_idx];
+            /* Find candidate elements with presence flag set and matching h2 hash. */
+            for (int pos = 0; pos < ELEMENTS_PER_BUCKET; pos++) {
+                if ((b->presence & (1 << pos)) && b->hashes[pos] == h2) {
+                    /* It's a candidate. */
+                    void *element = b->elements[pos];
+                    const void *elem_key = elementGetKey(s, element);
+                    if (compareKeys(s, key, elem_key) == 0) {
+                        /* It's a match. */
+                        if (pos_in_bucket) *pos_in_bucket = pos;
+                        if (table_index) *table_index = table;
+                        return b;
+                    }
+                }
+            }
+
+            /* Probe the next bucket? */
+            if (!b->everfull) break;
+            bucket_idx = nextCursor(bucket_idx, mask);
+            if (bucket_idx == start_bucket_idx) {
+                /* We probed the whole table. This should be extremely rare but
+                 * theoretically it can happen. */
+                break;
+            }
```

Comment:
This isn't protected against on the insertion path. I think we should make this impossible though, on the insertion we should always force the rehash long before the table is filled.

Reply:
Wherever we set the `everfull` flag on a bucket, in `insert`, we call `cleanUpTombstonesIfNeeded` which does starts rehashing if it's too full.

That wasn't implemented when I wrote this comment. We can delete this check or replace it with an assert.

Reply:
I tried replacing this with an assert, but then it asserts in a unit test case.

On the insert patch, we call cleanUpTombstonesIfNeeded *after* inserting the element, so it can happen that all buckets got full by that insert. It can only happen if filling one more bucket got us from below MAX_FILL_PERCENT_SOFT to 100% in one go, i.e. only for small tables with up to 4 buckets. In this case, rehashing is already in progress and we have two tables to check, where the second is not full.

I'm updating the comment but keeping this `if`.